src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpSender.php

```diff
@@ -47,20 +47,22 @@ public function send(Envelope $envelope): Envelope
         $delayStamp = $envelope->last(DelayStamp::class);
         $delay = $delayStamp ? $delayStamp->getDelay() : 0;
 
+        /** @var AmqpStamp|null $amqpStamp */
         $amqpStamp = $envelope->last(AmqpStamp::class);
         if (isset($encodedMessage['headers']['Content-Type'])) {
             $contentType = $encodedMessage['headers']['Content-Type'];
             unset($encodedMessage['headers']['Content-Type']);
 
-            $attributes = $amqpStamp ? $amqpStamp->getAttributes() : [];
-
-            if (!isset($attributes['content_type'])) {
-                $attributes['content_type'] = $contentType;
-
-                $amqpStamp = new AmqpStamp($amqpStamp ? $amqpStamp->getRoutingKey() : null, $amqpStamp ? $amqpStamp->getFlags() : AMQP_NOPARAM, $attributes);
+            if (!$amqpStamp || !isset($amqpStamp->getAttributes()['content_type'])) {
+                $amqpStamp = AmqpStamp::createWithAttributes(['content_type' => $contentType], $amqpStamp);
             }
         }
 
+        $amqpReceivedStamp = $envelope->last(AmqpReceivedStamp::class);
+        if ($amqpReceivedStamp instanceof AmqpReceivedStamp) {
+            $amqpStamp = AmqpStamp::createFromAmqpEnvelope($amqpReceivedStamp->getAmqpEnvelope(), $amqpStamp);
```

Comment:
Will this work when it's being sent to the failure transport... and the failure transport is a direct AMQP exchange that relies on the `default_publish_routing_key` config to route to some "failure" queue?

I think flow would be:

A) Message is received for last retry
B) Message fails again and so is sent to the failure transport.
C) This code will see the routingKey from the original message (e.g. `foo`) and set it on the `AmqpStamp`
D) When this code is forwarded into `Connection::publish()` below, the routing key (`foo`) in `AmqpStamp` will be used instead of the `default_publish_routing_key` from the "failure transport".
E) Message will be sent once again with the `foo` routing key, instead of the one from the failure transport.

If I'm correct, on a high level, it *seems* like we want to do this trick *unless* we detect that the message is being sent through a *different* transport than it was originally sent through.... or maybe the "failure transport" is added as an edge-case.

Reply:
What you described is also what is described in #32994 but the problem is that the failed transport should not need to be direct. I'll answer there.

Reply:
Documenting that a failure transport should not be direct is a valid solution I think. But:

> So retrying from the failure transport might not work if the routing key is different suddenly.

I don't think I understand what you're saying here. Assuming we successfully deliver a message to the routing transport (by setting the routing key correctly for the situation), what problem would the routing key being different cause? At that point, the routing key isn't used anymore: the `messenger:failed:retry` command consumes from the failure transport and dispatches into the bus. If it failed *again*, then re-using the "last routing key" (what this PR does) would correctly send it back to the failure transport, right?

Reply:
People could use the routing key using the AmqpReceivedStamp in their handler to make decisions. But again all these things are edge cases I would not worry about unless you have a certain change request in mind?

Reply:
Well, the options are:

1) document not to use a direct transport as your failure transport.

2) somehow add code here for the failure transport situation: like look for the RedeliveredToFailureTransportStamp... or something cleaner (not sure what). If there’s no reasonable way to do this, then option 1 becomes the only option.

Reply:
Honestly there is nothing to document or fix here. Your case is just one of endless configuration possibilities. I can also configure the failure transport with the appropriate binding keys and voila it still works with the original routing keys. We shouldn't enforce or document a random edge case here that does not apply always.

Reply:
What this PR is about is publishing retry/failure with the same routing key and priority etc as the original message. And this is absolutely the right and logical thing to do because retry message should be the same as the original ones.

Reply:
The default is a fanout exchange. So we're covering 99% of all cases. If someone changes it to a direct exchange for failure, it's not the messenger job to interfere with this and make certain assumptions.